platform/ios/src/MGLScaleBar.mm

```diff
@@ -0,0 +1,359 @@
+    _containerView.backgroundColor = self.secondaryColor;
+    [self addSubview:_containerView];
+    
+    _borderLayer = [CAShapeLayer layer];
+    _borderLayer.borderColor = [self.primaryColor CGColor];
+    _borderLayer.borderWidth = 1.0f / [[UIScreen mainScreen] scale];
+    
+    [_containerView.layer addSublayer:_borderLayer];
+    
+    _formatter = [[MGLDistanceFormatter alloc] init];
+}
+
+#pragma mark - Dimensions
+
+- (CGSize)intrinsicContentSize {
+    return CGSizeMake(self.actualWidth, 16);
+}
+
+- (CGFloat)actualWidth {
+    return self.row.distance / [self unitsPerPoint];
+}
+
+- (CGFloat)maximumWidth {
+    CGFloat fullWidth = CGRectGetWidth(self.superview.bounds);
+    CGFloat padding = [self usesRightToLeftLayout] ? fullWidth - CGRectGetMaxX(self.frame) : CGRectGetMinX(self.frame);
+    return floorf(fullWidth / 2 - padding);
+}
+
+- (CGFloat)unitsPerPoint {
+    return [self usesMetricSystem] ? self.metersPerPoint : self.metersPerPoint * MGLFeetPerMeter;
+}
+
+#pragma mark - Convenient methods
+
+- (BOOL)usesRightToLeftLayout {
+    return [UIView userInterfaceLayoutDirectionForSemanticContentAttribute:self.superview.semanticContentAttribute] == UIUserInterfaceLayoutDirectionRightToLeft;
+}
+
+- (BOOL)usesMetricSystem {
+    NSLocale *locale = [NSLocale currentLocale];
+    return [[locale objectForKey:NSLocaleUsesMetricSystem] boolValue];
+}
+
+- (MGLRow)preferredRow {
+    CLLocationDistance maximumDistance = [self maximumWidth] * [self unitsPerPoint];
+    MGLRow row;
+    
+    BOOL useMetric = [self usesMetricSystem];
+    NSUInteger count = useMetric
+    ? sizeof(MGLMetricTable) / sizeof(MGLMetricTable[0])
+    : sizeof(MGLImperialTable) / sizeof(MGLImperialTable[0]);
+    
+    for (NSUInteger i = 0; i < count; i++) {
+        CLLocationDistance distance = useMetric ? MGLMetricTable[i].distance : MGLImperialTable[i].distance;
+        if (distance <= maximumDistance) {
+            row = useMetric ? MGLMetricTable[i] : MGLImperialTable[i];
+        } else {
+            break;
+        }
+    }
+    
+    return row;
+}
+
+#pragma mark - Setters
+
+- (void)setMetersPerPoint:(CLLocationDistance)metersPerPoint {
+    if (_metersPerPoint == metersPerPoint) {
+        return;
+    }
+    
+    _metersPerPoint = metersPerPoint;
+    
+    [self updateVisibility];
+    
+    self.row = [self preferredRow];
+    
+    [self invalidateIntrinsicContentSize];
+    [self setNeedsLayout];
+}
+
+- (void)updateVisibility {
+    BOOL metric = [self usesMetricSystem];
+    
+    NSUInteger count = metric
+    ? sizeof(MGLMetricTable) / sizeof(MGLMetricTable[0])
+    : sizeof(MGLImperialTable) / sizeof(MGLImperialTable[0]);
+    
+    CLLocationDistance maximumDistance = [self maximumWidth] * [self unitsPerPoint];
+    CLLocationDistance allowedDistance = metric
+    ? MGLMetricTable[count-1].distance
+    : MGLImperialTable[count-1].distance;
+    
+    self.alpha = maximumDistance > allowedDistance ? 0 : 1;
```

Comment:
MapKit fades the scale bar in and out as it appears and disappears due to the maximum distance threshold. We could use a UIView animation block here perhaps.